services/payments/src/simcore_service_payments/api/rpc/_payments.py

```diff
@@ -86,3 +94,57 @@ async def init_payment(
         payment_id=f"{payment_id}",
         payment_form_url=f"{submission_link}",
     )
+
+
+@router.expose()
+async def cancel_payment(
+    app: FastAPI,
+    *,
+    payment_id: PaymentID,
+    user_id: UserID,
+    wallet_id: WalletID,
+) -> None:
+
+    # validation
+    repo = PaymentsTransactionsRepo(db_engine=app.state.engine)
+    payment = await repo.get_payment_transaction(
+        payment_id=payment_id, user_id=user_id, wallet_id=wallet_id
+    )
+
+    if payment is None:
+        raise PaymentNotFoundError(payment_id=payment_id)
+
+    if payment.state.is_completed():
+        if payment.state == PaymentTransactionState.CANCELED:
+            # Avoids error if multiple cancel calls
+            return
+        raise PaymentAlreadyAckedError(payment_id=payment_id)
+
+    # execution
+    with log_context(
+        _logger,
+        logging.INFO,
+        "%s: Cancelling payment %s in payments-gateway",
+        PAG,
+        f"{wallet_id=}",
+        extra=get_log_record_extra(user_id=user_id),
+    ):
+        payments_gateway_api = PaymentsGatewayApi.get_from_app_state(app)
+
+        init = PaymentInitiated(payment_id=payment_id)
+        payment_cancelled = await payments_gateway_api.cancel_payment(init)
+
+    with log_context(
+        _logger,
+        logging.INFO,
+        "%s: Annotate CANCEL transaction %s in db",
+        PGDB,
+        f"{payment_id=}",
+        extra=get_log_record_extra(user_id=user_id),
+    ):
+        await repo.update_ack_payment_transaction(
```

Comment:
stupid question here: what happens in the event where the DB is not accessible here? can the cancellation be retrieved from the gateway somehow?

Reply:
not stupid at all. Would you "annotate" it in the database before or after the gateway executes successfuly the cancelation? I think in both cases you have a chance of an error. 

I still need to make an extra round exclusively for error handling and how to recover undefined states.